File: software-dev-fullflow/skills/oil-oil/see-skill/scripts/ocr_macos.js
ObjC.import("CoreGraphics");
ObjC.import("Foundation");
ObjC.import("ImageIO");
ObjC.import("Vision");

function unwrap(value) {
  return ObjC.unwrap(value);
}

function run(argv) {
  if (argv.length !== 1) {
    throw new Error("Usage: ocr_macos.js <image-path>");
  }

  const path = argv[0];
  const url = $.NSURL.fileURLWithPath(path);
  const source = $.CGImageSourceCreateWithURL(url, null);
  if (!source) {
    throw new Error("Cannot decode image: " + path);
  }
  const image = $.CGImageSourceCreateImageAtIndex(source, 0, null);
  if (!image) {
    throw new Error("Cannot decode image: " + path);
  }

  const request = $.VNRecognizeTextRequest.alloc.init;
  request.recognitionLevel = 0;
  request.usesLanguageCorrection = true;
  if (request.respondsToSelector("supportedRecognitionLanguagesAndReturnError:")) {
    const languageError = Ref();
    const supported = request.supportedRecognitionLanguagesAndReturnError(languageError);
    const preferred = ["zh-Hans", "zh-Hant", "en-US"];
    const selected = [];
    for (let index = 0; index < preferred.length; index += 1) {
      if (supported.containsObject($(preferred[index]))) {
        selected.push(preferred[index]);
      }
    }
    if (selected.length > 0) {
      request.recognitionLanguages = $(selected);
    }
  }

  const handler = $.VNImageRequestHandler.alloc.initWithURLOptions(
    url,
    $.NSDictionary.dictionary
  );
  const error = Ref();
  const requests = $.NSArray.arrayWithObject(request);
  if (!handler.performRequestsError(requests, error)) {
    const detail = error[0] ? unwrap(error[0].localizedDescription) : "unknown error";
    throw new Error("Vision OCR failed: " + detail);
  }

  const items = [];
  const results = request.results;
  for (let index = 0; index < Number(results.count); index += 1) {
    const observation = results.objectAtIndex(index);
    const candidates = observation.topCandidates(1);
    if (Number(candidates.count) === 0) {
      continue;
    }
    const candidate = candidates.objectAtIndex(0);
    const box = observation.boundingBox;
    items.push({
      text: unwrap(candidate.string),
      confidence: Number(candidate.confidence),
      box: {
        x: Number(box.origin.x),
        y: Number(1 - box.origin.y - box.size.height),
        width: Number(box.size.width),
        height: Number(box.size.height),
      },
    });
  }

  return JSON.stringify({
    backend: "macos-vision",
    width: Number($.CGImageGetWidth(image)),
    height: Number($.CGImageGetHeight(image)),
    items,
  });
}

File: software-dev-fullflow/skills/oil-oil/see-skill/scripts/ocr_macos.swift
#!/usr/bin/env swift

import CoreGraphics
import Darwin
import Foundation
import ImageIO
import Vision

func fail(_ message: String) -> Never {
    FileHandle.standardError.write(Data(("[ERROR] " + message + "\n").utf8))
    exit(1)
}

func normalizedBox(_ box: CGRect) -> [String: Double] {
    return [
        "x": Double(box.minX),
        "y": Double(1.0 - box.maxY),
        "width": Double(box.width),
        "height": Double(box.height),
    ]
}

guard CommandLine.arguments.count == 2 else {
    fail("Usage: ocr_macos.swift <image-path>")
}

let imageURL = URL(fileURLWithPath: CommandLine.arguments[1])
guard let source = CGImageSourceCreateWithURL(imageURL as CFURL, nil),
      let image = CGImageSourceCreateImageAtIndex(source, 0, nil) else {
    fail("Cannot decode image: \(imageURL.path)")
}

let textRequest = VNRecognizeTextRequest()
textRequest.recognitionLevel = .accurate
textRequest.usesLanguageCorrection = true

if #available(macOS 13.0, *) {
    textRequest.automaticallyDetectsLanguage = true
}

do {
    let supported = try textRequest.supportedRecognitionLanguages()
    let preferred = ["zh-Hans", "zh-Hant", "en-US"]
    let selected = preferred.filter { supported.contains($0) }
    if !selected.isEmpty {
        textRequest.recognitionLanguages = selected
    }

    let handler = VNImageRequestHandler(cgImage: image, options: [:])
    try handler.perform([textRequest])

    let observations = (textRequest.results ?? []).sorted {
        let verticalDifference = abs($0.boundingBox.midY - $1.boundingBox.midY)
        if verticalDifference > 0.02 {
            return $0.boundingBox.midY > $1.boundingBox.midY
        }
        return $0.boundingBox.minX < $1.boundingBox.minX
    }

    let items: [[String: Any]] = observations.compactMap { observation in
        guard let candidate = observation.topCandidates(1).first else {
            return nil
        }
        let box = observation.boundingBox
        return [
            "text": candidate.string,
            "confidence": Double(candidate.confidence),
            "box": normalizedBox(box),
        ]
    }

    var sceneLabels: [[String: Any]] = []
    var people: [[String: Any]] = []
    var faces: [[String: Any]] = []
    var barcodes: [[String: Any]] = []
    var rectangles: [[String: Any]] = []
    var salientObjects: [[String: Any]] = []
    var contourCount = 0

    if #available(macOS 10.15, *) {
        let classifyRequest = VNClassifyImageRequest()
        if (try? handler.perform([classifyRequest])) != nil {
            sceneLabels = (classifyRequest.results ?? [])
                .filter { $0.confidence >= 0.10 }
                .sorted { $0.confidence > $1.confidence }
                .prefix(10)
                .map {
                    [
                        "identifier": $0.identifier,
                        "confidence": Double($0.confidence),
                    ]
                }
        }

        let humanRequest = VNDetectHumanRectanglesRequest()
        if (try? handler.perform([humanRequest])) != nil {
            people = (humanRequest.results ?? []).map {
                [
                    "confidence": Double($0.confidence),
                    "box": normalizedBox($0.boundingBox),
                ]
            }
        }

        let saliencyRequest = VNGenerateAttentionBasedSaliencyImageRequest()
        if (try? handler.perform([saliencyRequest])) != nil {
            salientObjects = (saliencyRequest.results?.first?.salientObjects ?? []).map {
                [
                    "confidence": Double($0.confidence),
                    "box": normalizedBox($0.boundingBox),
                ]
            }
        }
    }

    let faceRequest = VNDetectFaceRectanglesRequest()
    if (try? handler.perform([faceRequest])) != nil {
        faces = (faceRequest.results ?? []).map {
            [
                "confidence": Double($0.confidence),
                "box": normalizedBox($0.boundingBox),
            ]
        }
    }

    let barcodeRequest = VNDetectBarcodesRequest()
    if (try? handler.perform([barcodeRequest])) != nil {
        barcodes = (barcodeRequest.results ?? []).map {
            [
                "payload": $0.payloadStringValue ?? "",
                "symbology": $0.symbology.rawValue,
                "confidence": Double($0.confidence),
                "box": normalizedBox($0.boundingBox),
            ]
        }
    }

    let rectangleRequest = VNDetectRectanglesRequest()
    rectangleRequest.maximumObservations = 20
    rectangleRequest.minimumConfidence = 0.30
    rectangleRequest.minimumSize = 0.03
    if (try? handler.perform([rectangleRequest])) != nil {
        rectangles = (rectangleRequest.results ?? []).map {
            [
                "confidence": Double($0.confidence),
                "box": normalizedBox($0.boundingBox),
            ]
        }
    }

    if #available(macOS 11.0, *) {
        let contourRequest = VNDetectContoursRequest()
        if (try? handler.perform([contourRequest])) != nil {
            contourCount = contourRequest.results?.first?.topLevelContourCount ?? 0
        }
    }

    let result: [String: Any] = [
        "backend": "macos-vision-enhanced",
        "width": image.width,
        "height": image.height,
        "items": items,
        "scene_labels": sceneLabels,
        "people": people,
        "faces": faces,
        "barcodes": barcodes,
        "rectangles": rectangles,
        "salient_objects": salientObjects,
        "contour_count": contourCount,
    ]
    let data = try JSONSerialization.data(withJSONObject: result, options: [])
    FileHandle.standardOutput.write(data)
    FileHandle.standardOutput.write(Data("\n".utf8))
} catch {
    fail("Vision OCR failed: \(error.localizedDescription)")
}
